### PR TITLE
remove type coercion in the binary physical expr

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3388

depend on https://github.com/apache/arrow-datafusion/issues/3421 https://github.com/apache/arrow-datafusion/issues/3289 https://github.com/apache/arrow-datafusion/pull/3459 https://github.com/apache/arrow-datafusion/pull/3472

fixed the row filter predication for the data type of `NULL` value https://github.com/apache/arrow-datafusion/pull/3470

 # Rationale for this change

# What changes are included in this PR?

In order to make the pr review more friendly for reviewer, I will explain what I did in this pr:

1.   remove the type coercion in the creation of the [binary physical expr](https://github.com/apache/arrow-datafusion/blob/3062cd5d73104a0d504cff518ec19544d2beeb75/datafusion/physical-expr/src/expressions/binary.rs) and fix the test case

2. redo the `simplify expression` again after the `type coercion` in the [optimizer](https://github.com/apache/arrow-datafusion/blob/3062cd5d73104a0d504cff518ec19544d2beeb75/datafusion/core/src/execution/context.rs#L1459). This is just an minor fix. If the issue https://github.com/apache/arrow-datafusion/issues/3582 done, we can remove the duplicated `simplify expression` in the optimizer.

3. try do the  type coercion before evaluate the logical expr [in the file](https://github.com/apache/arrow-datafusion/blob/3062cd5d73104a0d504cff518ec19544d2beeb75/datafusion/optimizer/src/simplify_expressions.rs#L504)  according to the comments in this issue https://github.com/apache/arrow-datafusion/issues/3556 from @alamb . This is just an temporary fix, and it will be remove when do https://github.com/apache/arrow-datafusion/issues/3582

4. remove the between optimization in the [simplify expression](https://github.com/apache/arrow-datafusion/blob/3062cd5d73104a0d504cff518ec19544d2beeb75/datafusion/optimizer/src/simplify_expressions.rs#L818) because of issue https://github.com/apache/arrow-datafusion/issues/3587, and will recovery when do https://github.com/apache/arrow-datafusion/issues/3582

5. refresh the test cases

cc @alamb @andygrove 

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -917,8 +898,16 @@ pub fn binary(
...
-    Ok(Arc::new(BinaryExpr::new(l, op, r)))
+    let lhs_type = &lhs.data_type(input_schema)?;
+    let rhs_type = &rhs.data_type(input_schema)?;
+    if !lhs_type.eq(rhs_type) {
+        return Err(DataFusionError::Internal(format!(
+            "The left type {} and right type {} of binary physical expr should be same",
+            lhs_type, rhs_type
+        )));
+    }
+    // let (l, r) = binary_cast(lhs, &op, rhs, input_schema)?;
```
Comment (andygrove) [inline datafusion/physical-expr/src/expressions/binary.rs:909]:
```suggestion
```

**Comment by liukun4515** [comment]
after rebase https://github.com/apache/arrow-datafusion/pull/3380, the test case `evolved_schema_intersection_filter_with_filter_pushdown` and `evolved_schema_disjoint_schema_filter_with_pushdown` failed

**Comment by liukun4515** [comment]
this pr depend on the https://github.com/apache/arrow-datafusion/pull/3470 and https://github.com/apache/arrow-datafusion/pull/3472

**Comment by liukun4515** [comment]
I meet so many issue when remove the binary type coercion which is caused from these pr https://github.com/apache/arrow-datafusion/pull/3301 https://github.com/apache/arrow-datafusion/pull/3246
For example  `is TRUE`, these is a logical expr `Expr::IsTrue` but convert it to the physical binary op in https://github.com/andygrove/arrow-datafusion/blob/92b1a341aca334c54235ef65737c88037010e30b/datafusion/physical-expr/src/planner.rs#L86

@andygrove  why not generate the logical binary op when create the `Expr::IsTrue` logical expr.

**Comment by liukun4515** [comment]
> I meet so many issue when remove the binary type coercion which is caused from these pr #3301 #3246 For example `is TRUE`, these is a logical expr `Expr::IsTrue` but convert it to the physical binary op in https://github.com/andygrove/arrow-datafusion/blob/92b1a341aca334c54235ef65737c88037010e30b/datafusion/physical-expr/src/planner.rs#L86
> 
> @andygrove why not generate the logical binary op when create the `Expr::IsTrue` logical expr.

cc @alamb 

I create the issue https://github.com/apache/arrow-datafusion/issues/3509 to support other logical expr for the typ coercion.
If we can convert such expr like `istrue, isfalse,like` to `binary op` in the logical phase, we don't need to do this by adding more case in the TypeCoercion optimizer rule.
I don't why we don't do this when implement the `is true, is false` in the https://github.com/apache/arrow-datafusion/pull/3301/files @andygrove , is there any optimization should be done in the logical phase with the `Expr:IsTrue`?

**Comment by alamb** [comment]
>  is there any optimization should be done in the logical phase with the Expr:IsTrue?

I think the only coercion it might need is that its argument is `DataType::Boolean`

**Comment by andygrove** [comment]
I've started reviewing this, but there are many changes and referenced issues I need to go read. I will continue with the review tomorrow.

**Comment by alamb** [comment]
I also plan to review this PR, but I may not get to it today

**Comment by liukun4515** [comment]
I try to explain what I do and why i do in the description `What changes are included in this PR?`

I hope this works for your review.

cc @andygrove @alamb

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1452,7 +1452,11 @@ impl SessionState {
             rules.push(Arc::new(FilterNullJoinKeys::default()));
         }
         rules.push(Arc::new(ReduceOuterJoin::new()));
+        // TODO: https://github.com/apache/arrow-datafusion/issues/3557
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1455]: 👍

Comment (alamb) [reply datafusion/core/src/execution/context.rs:1455]: it makes sense to me that we need to simplify expressons after coercion

File: datafusion/core/tests/sql/aggregates.rs [hunk 3]
```diff
@@ -1834,11 +1834,11 @@ async fn aggregate_avg_add() -> Result<()> {
     assert_eq!(results.len(), 1);
 
     let expected = vec![
-        "+--------------+---------------------------+---------------------------+---------------------------+",
-        "| AVG(test.c1) | AVG(test.c1) + Float64(1) | AVG(test.c1) + Float64(2) | Float64(1) + AVG(test.c1) |",
-        "+--------------+---------------------------+---------------------------+---------------------------+",
-        "| 1.5          | 2.5                       | 3.5                       | 2.5                       |",
-        "+--------------+---------------------------+---------------------------+---------------------------+",
+        "+--------------+-------------------------+-------------------------+-------------------------+",
```
Comment (alamb) [inline datafusion/core/tests/sql/aggregates.rs:1837]: 👍

File: datafusion/expr/src/binary_rule.rs [hunk 4]
```diff
@@ -801,4 +804,230 @@ mod tests {
...
+            Operator::Divide,
+            DataType::Decimal128(23, 11)
+        );
+        test_coercion_binary_rule!(
+            DataType::Int32,
+            DataType::Decimal128(10, 2),
+            Operator::Modulo,
+            DataType::Decimal128(10, 2)
+        );
+        // TODO add other data type
```
Comment (alamb) [inline datafusion/expr/src/binary_rule.rs:934]: I wonder if we should file another ticket to track this gap? Specifically, a "help wanted" ticket that explained what was needed might encourage some additional contributions

File: datafusion/expr/src/binary_rule.rs [hunk 5]
```diff
@@ -801,4 +804,230 @@ mod tests {
...
+            Operator::Eq,
+            DataType::Decimal128(20, 0)
+        );
+        test_coercion_binary_rule!(
+            DataType::Decimal128(14, 2),
+            DataType::Decimal128(10, 3),
+            Operator::GtEq,
+            DataType::Decimal128(15, 3)
+        );
+
```
Comment (alamb) [inline datafusion/expr/src/binary_rule.rs:1011]: I reviewed the tests cases in this file carefully. Very nice 👌

File: datafusion/expr/src/binary_rule.rs [hunk 6]
```diff
@@ -801,4 +804,230 @@ mod tests {
...
+            DataType::Boolean,
+            Operator::And,
+            DataType::Boolean
+        );
+
+        test_coercion_binary_rule!(
+            DataType::Boolean,
+            DataType::Boolean,
+            Operator::Or,
+            DataType::Boolean
```
Comment (alamb) [inline datafusion/expr/src/binary_rule.rs:1029]: Is there any other types that are coerced to boolean for logical operations? Or are the tests for boolean just showing that boolean types are not changed when coerced?

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 7]
```diff
@@ -360,6 +362,7 @@ pub struct ConstEvaluator<'a> {
     execution_props: &'a ExecutionProps,
     input_schema: DFSchema,
     input_batch: RecordBatch,
+    type_coercion_helper: TypeCoercionRewriter,
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:365]:
```suggestion
    // Needed until we ensure type coercion is done before any optimizations
    // https://github.com/apache/arrow-datafusion/issues/3557
    type_coercion_helper: TypeCoercionRewriter,
```

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 8]
```diff
@@ -804,24 +817,28 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
             //
             // Rules for Between
             //
+            // TODO https://github.com/apache/arrow-datafusion/issues/3587
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:820]: makes sense -- thank you for the comments

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 9]
```diff
@@ -1107,6 +1124,12 @@ mod tests {
         assert_eq!(simplify(expr_eq), lit(true));
     }
 
+    #[test]
+    fn test_simplify_with_type_coercion() {
+        let expr_plus = binary_expr(lit(1_i32), Operator::Plus, lit(1_i64));
+        assert_eq!(simplify(expr_plus), lit(2_i64));
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1130]: Nice

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 10]
```diff
@@ -1573,13 +1597,11 @@ mod tests {
...
         };
+        let between_expr = expr.clone();
         let expr = expr.or(lit_bool_null());
         let result = simplify(expr);
 
-        let expected_expr = or(
-            and(col("c1").gt_eq(lit(0)), col("c1").lt_eq(lit(10))),
-            lit_bool_null(),
-        );
+        let expected_expr = or(between_expr, lit_bool_null());
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1604]:
```suggestion
        let expected_expr = or(between_expr, lit_bool_null());
        // let expected_expr = or(
        //    and(col("c1").gt_eq(lit(0)), col("c1").lt_eq(lit(10))),
        //    lit_bool_null(),
        //);        
```

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 11]
```diff
@@ -1611,17 +1635,17 @@ mod tests {
...
         };
+        let between_expr = expr.clone();
         let expr = expr.and(lit_bool_null());
         let result = simplify(expr);
 
-        let expected_expr = and(
-            and(col("c1").gt_eq(lit(0)), col("c1").lt_eq(lit(10))),
-            lit_bool_null(),
-        );
+        let expected_expr = and(between_expr, lit_bool_null());
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1642]:
```suggestion
        let expected_expr = and(between_expr, lit_bool_null());
        // let expected_expr = and(
        //    and(col("c1").gt_eq(lit(0)), col("c1").lt_eq(lit(10))),
        //    lit_bool_null(),
        // );        
```

File: datafusion/optimizer/src/simplify_expressions.rs (hunk 10 above)
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions.rs:1604]: I suggest keeping the old expected result in the code as comments to make it easier to understand what the desired effect is

File: datafusion/optimizer/src/type_coercion.rs [hunk 12]
```diff
@@ -69,14 +67,8 @@ impl OptimizerRule for TypeCoercion {
             },
         );
 
-        let mut execution_props = ExecutionProps::new();
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:72]: Very nice

File: datafusion/optimizer/tests/integration-test.rs [hunk 13]
```diff
@@ -79,31 +79,34 @@ fn intersect() -> Result<()> {
...
         "Projection: #COUNT(UInt8(1))\n  Aggregate: groupBy=[[]], aggr=[[COUNT(UInt8(1))]]\
-        \n    Filter: #test.col_date64 >= Date64(\"890179200000\") AND #test.col_date64 <= Date64(\"897955200000\")\
+        \n    Filter: #test.col_date64 BETWEEN Date64(\"890179200000\") AND Date64(\"897955200000\")\
         \n      TableScan: test projection=[col_date64]";
     assert_eq!(expected, format!("{:?}", plan));
     Ok(())
 }
 
 fn test_sql(sql: &str) -> Result<LogicalPlan> {
+    // TODO should make align with rules in the context
```
Comment (alamb) [inline datafusion/optimizer/tests/integration-test.rs:109]:
```suggestion
    // TODO should make align with rules in the context
    // https://github.com/apache/arrow-datafusion/issues/3524
```

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 14]
```diff
@@ -892,25 +891,6 @@ impl BinaryExpr {
     }
 }
 
-/// return two physical expressions that are optionally coerced to a
-/// common type that the binary operator supports.
-fn binary_cast(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:897]: 🎉

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 15]
```diff
@@ -1080,7 +1071,7 @@ mod tests {
             Operator::Plus,
             Int32Array,
             DataType::Int32,
-            vec![2i32, 4i32]
+            vec![2i32, 4i32],
         );
         test_coercion!(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:1076]: I wonder if these tests are doing anything useful anymore now that we are coercing in the logical layer? 🤔 It seems like they are now testing the test code 😆

**Review by alamb — APPROVED** [review]
This was a great PR to read @liukun4515  -- really nice. 

Thank you for sticking with it. 

I think we should merge it asap to minimize conflicts.

**Comment by alamb** [comment]
I plan to get this PR ready to merge by: add my suggested comments, fix clippy, and merge to master

**Comment by alamb** [comment]
(since I already have it checked out this will be a simple thing for me)

**Pull request merged**